cmd/ipfs/daemon.go

```diff
@@ -288,8 +292,36 @@ func daemonFunc(req *cmds.Request, re cmds.ResponseEmitter, env cmds.Environment
 			return fmt.Errorf("fs-repo requires migration")
 		}
 
-		// Fetch migrations from current distribution, or location from environ
-		fetcher := migrations.NewHttpFetcher(migrations.GetDistPathEnv(migrations.CurrentIpfsDist), "", "go-ipfs", 0)
+		migrationCfg, err := readMigrationConfig(cctx.ConfigRoot)
+		if err != nil {
+			return err
+		}
+
+		fetcher, err = getMigrationFetcher(migrationCfg, &cctx.ConfigRoot)
+		if err != nil {
+			return err
+		}
+		defer fetcher.Close()
+
+		if migrationCfg.Keep == "cache" {
+			cacheMigrations = true
+		} else if migrationCfg.Keep == "pin" {
+			pinMigrations = true
+		}
+
+		if cacheMigrations || pinMigrations {
+			// Create temp directory to store downloaded migration archives
+			migrations.DownloadDirectory, err = ioutil.TempDir("", "migrations")
+			if err != nil {
+				return err
+			}
+			defer func() {
+				if migrations.DownloadDirectory != "" {
+					os.RemoveAll(migrations.DownloadDirectory)
+				}
+			}()
```

Comment:
This is done in a deferred function here, so that cleanup will happen if the daemon returns early due to an error.  It is done later after finishing migration work, so that that directory is not sitting around for the lifetime of the daemon and is not left behind if the daemon exits ungracefully.

Added comment to explain.